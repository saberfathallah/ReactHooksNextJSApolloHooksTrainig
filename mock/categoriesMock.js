export const CATEGORY = {
    id: "5e89c5f78245cd0d75dec50a",
    level: 1,
    name: "Sport",
    children: [
      {
        children: [],
        id: "5e89c6068245cd0d75dec50c",
        name: "foot",
        level: 2,
      },
    ],
  }