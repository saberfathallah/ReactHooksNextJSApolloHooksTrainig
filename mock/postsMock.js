export const POST = {
  categoryId: "5e89c6458245cd0d75dec50f",
  postId: "5e89c6458245cd0d75dec50f",
  comments: [],
  creatorId: "5e89c2fd8245cd0d75dec501",
  description: "corona france post",
  likes: ["5e8a26784c6b1d234b4c1a3f", "5e89c2fd8245cd0d75dec501"],
  id: "5e89c6878245cd0d75dec510",
  userId: {
    id: "5fac3671dd9b1b079249d522",
    email: "sberrr@gmail.com",
    name: "sberrr",
  },
};

// interface Post {
//   userId: User;
//   categoryId: string;
//   postId: string;
//   comments: Comment[];
//   description: string;
//   id: string;
//   likes: string[];
// }

// categoryId: "5e89c6458245cd0d75dec50f"
// comments: (5) [{…}, {…}, {…}, {…}, {…}]
// description: "corona france post"
// id: "5e89c6878245cd0d75dec510"
// likes: (2) ["5e8a26784c6b1d234b4c1a3f", "5e89c2fd8245cd0d75dec501"]
// userId: {name: "Saber Fathallah", id: "5e89c2fd8245cd0d75dec501", __typename: "User"}
