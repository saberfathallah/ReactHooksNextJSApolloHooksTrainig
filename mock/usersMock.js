export const USER_CONNECTED = {
    id: "5fac3671dd9b1b079249d522",
    email: "sberrr@gmail.com",
    name: "sberrr",
  };