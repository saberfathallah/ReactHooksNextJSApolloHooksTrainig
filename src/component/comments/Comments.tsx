import React from 'react';
import { useMutation, useQuery } from '@apollo/react-hooks';
import { Formik } from 'formik';
import TextField from '@material-ui/core/TextField';
import { createStyles, Theme, makeStyles } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import Button from '@material-ui/core/Button';
import Icon from '@material-ui/core/Icon';

import ADD_COMMENT, { updateCacheAfterAddComment } from '../../graphql/comments/mutation/addComment';
import GET_CURRENT_CATEGORY_ID from '../../graphql/client/queries/getCurrentCategoryId';
import Comment from '../comment';

const useStyles = makeStyles((theme: Theme) => createStyles({
  text: {
    padding: theme.spacing(2, 2, 0),
  },
  paper: {
    paddingBottom: 50,
  },
  list: {
    marginBottom: theme.spacing(2),
  },
  subheader: {
    backgroundColor: theme.palette.background.paper,
  },
  appBar: {
    top: 'auto',
    bottom: 0,
  },
  grow: {
    flexGrow: 1,
  },
  fabButton: {
    position: 'absolute',
    zIndex: 1,
    top: -30,
    left: 0,
    right: 0,
    margin: '0 auto',
  },
  textField: {
    marginLeft: theme.spacing(1),
    marginRight: theme.spacing(1),
    width: '25ch',
  },
  button: {
    margin: theme.spacing(1),
    float: 'right',
  },
}));

interface User {
  name: string;
  id: string;
}

interface Comment {
  description: string;
  userId: User;
  postId: string;
  id: string;
}
interface CommentProps {
  comments: Comment[];
  postId: string;
  categoryId: string;
  userConnected: User;
  creatorId: string;
}

interface CommnetFormInput {
  description: string;
  postId: string;
  categoryId: string;
}

const Comments: React.FC<CommentProps> = (props: CommentProps) => {
  const {
    comments, categoryId, postId, userConnected, creatorId,
  } = props;
  const { data: { currentCategoryId } } = useQuery(GET_CURRENT_CATEGORY_ID);
  const [addComment] = useMutation<any>(ADD_COMMENT,
    { update: (cache, { data }) => updateCacheAfterAddComment(cache, data, currentCategoryId) });
  const classes = useStyles();
  const initialValues: CommnetFormInput = { description: '', postId: '', categoryId: '' };

  return (
    <>
      <List className={classes.list}>
        {comments.map(({
          userId: { name, id: userCommentedId }, description, postId: postIdent, id,
        }) => (
          <React.Fragment key={description}>
            <ListItem button>
              <Comment
                currentCategoryId={currentCategoryId}
                id={id}
                postId={postIdent}
                creatorId={creatorId}
                userConnected={userConnected}
                name={name}
                description={description}
                userCommentedId={userCommentedId}
              />
            </ListItem>
          </React.Fragment>
        ))}
      </List>
      <Formik
        initialValues={initialValues}
        validate={(values): object => {
          let errors = {};
          if (!values.description) {
            errors = {
              ...errors,
              email: 'Required',
            };
          }
          return errors;
        }}
        onSubmit={async (values): Promise<void> => {
          await addComment({
            variables: {
              commentInput: {
                ...values,
                categoryId,
                postId,
              },
            },
          });
        }}
      >
        {({
          values,
          errors,
          touched,
          handleChange,
          handleBlur,
          handleSubmit,
        }): any => (
          <form onSubmit={handleSubmit}>
            <TextField
              type="description"
              name="description"
              onChange={handleChange}
              onBlur={handleBlur}
              value={values.description}
              id="standard-full-width"
              label="Ajouter un commentaire"
              style={{ margin: 8 }}
              placeholder=""
              fullWidth
              margin="normal"
              InputLabelProps={{
                shrink: true,
              }}
            />
            {errors.description && touched.description && errors.description}
            <Button
              type="submit"
              variant="contained"
              color="primary"
              size="small"
              className={classes.button}
              endIcon={<Icon>send</Icon>}
            >
              Commenter
            </Button>
          </form>
        )}
      </Formik>
    </>
  );
};

export default Comments;
