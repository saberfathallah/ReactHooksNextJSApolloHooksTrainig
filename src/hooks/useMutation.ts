import { useMutation as useMutationApollo } from "@apollo/react-hooks";

export const useMutation = (
  mutation: any,
  { variables, optionCache, ...otherOptions }: any
): any => {
  if (optionCache) {
    const [mutationFunction, { data, error, loading }] = useMutationApollo(
      mutation,
      {
        variables,
        update: (cache, { data }) =>
          optionCache.updateCacheFunction(cache, data, optionCache.variables),
          ...otherOptions,
      }
    );
    return [mutationFunction, { data, error, loading }];
  }

  const [mutationFunction, { data, error, loading }] = useMutationApollo(
    mutation,
    {
      variables,
    }
  );

  return [mutationFunction, { data, error, loading }];
};
