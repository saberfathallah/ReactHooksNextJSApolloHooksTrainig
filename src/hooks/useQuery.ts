import { useQuery as useQueryApollo } from "@apollo/react-hooks";

export const useQuery = (query, options={}): any => {
  const { loading, data, ...rest } = useQueryApollo(query, {
    ...options,
  });

  return { data, loading, ...rest };
}

// export default useQuery;

