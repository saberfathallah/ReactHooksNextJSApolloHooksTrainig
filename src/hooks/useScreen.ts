import { useEffect, useState } from "react";
import { isMobile, isMobileWidth } from "@utils/isBrowser";
import { ESKTOP_BREAKPOINT } from "@constants/dimensions";

const useScreen = (isClient) => {
  const [screen, setScreen] = useState({
    screenWidth: ESKTOP_BREAKPOINT,
    isMobileSize: isMobile,
  });

  useEffect(() => {
    if (!isClient) {
      return;
    }

    // first load
    setScreen({
      screenWidth: window.innerWidth,
      isMobileSize: isMobileWidth(window.innerWidth),
    });

    const handleResize = () => {
      if (isClient) {
        setScreen({
          screenWidth: window.innerWidth,
          isMobileSize: isMobileWidth(window.innerWidth),
        });
      }
    };

    window.addEventListener("resize", handleResize);

    return () => window.removeEventListener("resize", handleResize);
  }, [isClient]);

  return screen;
};

export default useScreen;
