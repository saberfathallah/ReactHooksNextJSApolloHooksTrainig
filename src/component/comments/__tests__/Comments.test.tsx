import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import Comments from "../Comments";
import { useMutation } from "../../../hooks/useMutation";
import { USER_CONNECTED } from '../../../../mock/usersMock';

jest.mock("../../../hooks/useMutation", () => ({
  useMutation: jest.fn(),
}));

// @ts-ignore
useMutation.mockImplementation(() => [() => {}, { loading: false }]);

describe("Comments", () => {
  const addComment = jest.fn();
  const props = {
    categoryId: "5e89c6458245cd0d75dec50f",
    comments: [
      {
        description: "comment bascket 2",
        id: "5e8e20ceb577a935e9c4cb45",
        postId: "5e8b5198780fc019e6d33595",
        userId: USER_CONNECTED,
      },
    ],
    creatorId: "5e89c2fd8245cd0d75dec501",
    postId: "5e89c6878245cd0d75dec510",
    userConnected: USER_CONNECTED,
    addComment,
    currentCategoryId:"6e89c6878245cd0d75dec510"
  }
  it("should display comments snapshot", () => {
    const tree = renderer.create(<Comments {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it("should display comments", () => {
    const tree = mount(<Comments {...props} />);
    expect(tree.text()).toContain("comment bascket 2");
  });

  // it("should call addComment", () => {
  //   const wrapper = mount(<Comments {...props} />);
  //   wrapper.find("button").simulate("click");
  //   expect(addComment).toHaveBeenCalledTimes(1);
  // });
});
