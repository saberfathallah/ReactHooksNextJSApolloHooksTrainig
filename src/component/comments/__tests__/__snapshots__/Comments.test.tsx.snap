// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments should display comments snapshot 1`] = `
Array [
  <ul
    className="MuiList-root makeStyles-list-3 MuiList-padding"
  >
    <div
      aria-disabled={false}
      className="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        id="comment-actions"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "display": "contents",
          }
        }
      >
        <div
          className="MuiListItemAvatar-root"
        >
          <div
            className="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root MuiAvatar-fallback"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
              />
            </svg>
          </div>
        </div>
        <div
          className="MuiListItemText-root MuiListItemText-multiline"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            sberrr
          </span>
          <p
            className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
          >
            comment bascket 2
          </p>
        </div>
      </div>
    </div>
  </ul>,
  <form
    onSubmit={[Function]}
  >
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      style={
        Object {
          "margin": 8,
        }
      }
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
        data-shrink={true}
        htmlFor="5e89c6878245cd0d75dec510standard-full-width"
        id="5e89c6878245cd0d75dec510standard-full-width-label"
      >
        Ajouter un commentaire
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          id="5e89c6878245cd0d75dec510standard-full-width"
          name="description"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          required={false}
          type="description"
          value=""
        />
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-9 MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Commenter
        <span
          className="MuiButton-endIcon MuiButton-iconSizeSmall"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root"
          >
            send
          </span>
        </span>
      </span>
    </button>
  </form>,
]
`;
